File: FLM_SubconLabelSystem/App_Code/GlobalFunctions.vb
Imports Microsoft.VisualBasic

Public Class GlobalFunctions

    Public Shared Function encrypt(ByVal plaintext As String) As String
        Dim AES As New System.Security.Cryptography.RijndaelManaged
        Dim SHA256 As New System.Security.Cryptography.SHA256Cng
        Dim ciphertext As String = ""
        Try
            AES.GenerateIV()
            AES.Key = SHA256.ComputeHash(System.Text.ASCIIEncoding.ASCII.GetBytes("N882A66A56299823AA026E8GARR1B527WDYY3A6EA2FF14F8DB70AB2829B69DDF"))

            AES.Mode = System.Security.Cryptography.CipherMode.CBC
            Dim DESEncrypter As System.Security.Cryptography.ICryptoTransform = AES.CreateEncryptor
            Dim Buffer As Byte() = System.Text.ASCIIEncoding.ASCII.GetBytes(plaintext)
            ciphertext = Convert.ToBase64String(DESEncrypter.TransformFinalBlock(Buffer, 0, Buffer.Length))

            Return Convert.ToBase64String(AES.IV) & Convert.ToBase64String(DESEncrypter.TransformFinalBlock(Buffer, 0, Buffer.Length))

        Catch ex As Exception
            Return ex.Message
        End Try
    End Function

    Public Shared Function decrypt(ByVal ciphertext As String) As String
        Dim AES As New System.Security.Cryptography.RijndaelManaged
        Dim SHA256 As New System.Security.Cryptography.SHA256Cng
        Dim plaintext As String = ""
        Dim iv As String = ""
        Try
            Dim delim As String() = New String(0) {"=="}
            Dim ivct = ciphertext.Split(delim, StringSplitOptions.None)
            iv = ivct(0) & "=="
            ciphertext = If(ivct.Length = 3, ivct(1) & "==", ivct(1))


            AES.Key = SHA256.ComputeHash(System.Text.ASCIIEncoding.ASCII.GetBytes("N882A66A56299823AA026E8GARR1B527WDYY3A6EA2FF14F8DB70AB2829B69DDF"))
            AES.IV = Convert.FromBase64String(iv)
            AES.Mode = System.Security.Cryptography.CipherMode.CBC
            Dim DESDecrypter As System.Security.Cryptography.ICryptoTransform = AES.CreateDecryptor
            Dim Buffer As Byte() = Convert.FromBase64String(ciphertext)
            plaintext = System.Text.ASCIIEncoding.ASCII.GetString(DESDecrypter.TransformFinalBlock(Buffer, 0, Buffer.Length))
            Return plaintext
        Catch ex As Exception
            Return ex.Message
        End Try
    End Function


    Public Shared Function retriveIV(ByVal ciphertext As String) As String
        Dim AES As New System.Security.Cryptography.RijndaelManaged
        Dim SHA256 As New System.Security.Cryptography.SHA256Cng
        Dim plaintext As String = ""
        Dim iv As String = ""
        Try
            Dim delim As String() = New String(0) {"=="}
            Dim ivct = ciphertext.Split(delim, StringSplitOptions.None)
            iv = ivct(0) & "=="
            ciphertext = If(ivct.Length = 3, ivct(1) & "==", ivct(1))


            AES.Key = SHA256.ComputeHash(System.Text.ASCIIEncoding.ASCII.GetBytes("N882A66A56299823AA026E8GARR1B527WDYY3A6EA2FF14F8DB70AB2829B69DDF"))
            AES.IV = Convert.FromBase64String(iv)
            AES.Mode = System.Security.Cryptography.CipherMode.CBC
            Dim DESDecrypter As System.Security.Cryptography.ICryptoTransform = AES.CreateDecryptor
            Dim Buffer As Byte() = Convert.FromBase64String(ciphertext)
            plaintext = System.Text.ASCIIEncoding.ASCII.GetString(DESDecrypter.TransformFinalBlock(Buffer, 0, Buffer.Length))
            Return iv
        Catch ex As Exception
            Return ex.Message
        End Try
    End Function

    Public Shared Function encryptIV(ByVal plaintext As String, ByVal iv As String) As String
        Dim AES As New System.Security.Cryptography.RijndaelManaged
        Dim SHA256 As New System.Security.Cryptography.SHA256Cng
        Dim ciphertext As String = ""
        Try

            AES.Key = SHA256.ComputeHash(System.Text.ASCIIEncoding.ASCII.GetBytes("N882A66A56299823AA026E8GARR1B527WDYY3A6EA2FF14F8DB70AB2829B69DDF"))
            AES.IV = Convert.FromBase64String(iv)
            AES.Mode = System.Security.Cryptography.CipherMode.CBC
            Dim DESEncrypter As System.Security.Cryptography.ICryptoTransform = AES.CreateEncryptor
            Dim Buffer As Byte() = System.Text.ASCIIEncoding.ASCII.GetBytes(plaintext)
            ciphertext = Convert.ToBase64String(DESEncrypter.TransformFinalBlock(Buffer, 0, Buffer.Length))

            Return Convert.ToBase64String(AES.IV) & Convert.ToBase64String(DESEncrypter.TransformFinalBlock(Buffer, 0, Buffer.Length))

        Catch ex As Exception
            Return ex.Message
        End Try
    End Function

End Class

File: FLM_SubconLabelSystem/Code/GlobalFunctions.cs
using System;
using System.Security.Cryptography;
using System.Text;

public class GlobalFunctions
{
    private const string SecretKey = "N882A66A56299823AA026E8GARR1B527WDYY3A6EA2FF14F8DB70AB2829B69DDF";

    public static string Encrypt(string plaintext)
    {
        using (var aes = Aes.Create())
        using (var sha256 = SHA256.Create())
        {
            try
            {
                aes.GenerateIV();
                aes.Key = sha256.ComputeHash(Encoding.ASCII.GetBytes(SecretKey));
                aes.Mode = CipherMode.CBC;

                ICryptoTransform encryptor = aes.CreateEncryptor();
                byte[] buffer = Encoding.ASCII.GetBytes(plaintext);

                return Convert.ToBase64String(aes.IV) + Convert.ToBase64String(encryptor.TransformFinalBlock(buffer, 0, buffer.Length));
            }
            catch (Exception ex)
            {
                return ex.Message;
            }
        }
    }

    public static string Decrypt(string ciphertext)
    {
        using (var aes = Aes.Create())
        using (var sha256 = SHA256.Create())
        {
            try
            {
                string[] delim = new string[] { "==" };
                string[] ivct = ciphertext.Split(delim, StringSplitOptions.None);
                string iv = ivct[0] + "==";
                ciphertext = ivct.Length == 3 ? ivct[1] + "==" : ivct[1];

                aes.Key = sha256.ComputeHash(Encoding.ASCII.GetBytes(SecretKey));
                aes.IV = Convert.FromBase64String(iv);
                aes.Mode = CipherMode.CBC;

                ICryptoTransform decryptor = aes.CreateDecryptor();
                byte[] buffer = Convert.FromBase64String(ciphertext);

                return Encoding.ASCII.GetString(decryptor.TransformFinalBlock(buffer, 0, buffer.Length));
            }
            catch (Exception ex)
            {
                return ex.Message;
            }
        }
    }

    public static string RetrieveIV(string ciphertext)
    {
        using (var aes = Aes.Create())
        using (var sha256 = SHA256.Create())
        {
            try
            {
                string[] delim = new string[] { "==" };
                string[] ivct = ciphertext.Split(delim, StringSplitOptions.None);
                string iv = ivct[0] + "==";
                ciphertext = ivct.Length == 3 ? ivct[1] + "==" : ivct[1];

                aes.Key = sha256.ComputeHash(Encoding.ASCII.GetBytes(SecretKey));
                aes.IV = Convert.FromBase64String(iv);
                aes.Mode = CipherMode.CBC;

                ICryptoTransform decryptor = aes.CreateDecryptor();
                byte[] buffer = Convert.FromBase64String(ciphertext);

                // Decryption performed to validate, but only IV is returned
                Encoding.ASCII.GetString(decryptor.TransformFinalBlock(buffer, 0, buffer.Length));

                return iv;
            }
            catch (Exception ex)
            {
                return ex.Message;
            }
        }
    }

    public static string EncryptIV(string plaintext, string iv)
    {
        using (var aes = Aes.Create())
        using (var sha256 = SHA256.Create())
        {
            try
            {
                aes.Key = sha256.ComputeHash(Encoding.ASCII.GetBytes(SecretKey));
                aes.IV = Convert.FromBase64String(iv);
                aes.Mode = CipherMode.CBC;

                ICryptoTransform encryptor = aes.CreateEncryptor();
                byte[] buffer = Encoding.ASCII.GetBytes(plaintext);

                return Convert.ToBase64String(aes.IV) + Convert.ToBase64String(encryptor.TransformFinalBlock(buffer, 0, buffer.Length));
            }
            catch (Exception ex)
            {
                return ex.Message;
            }
        }
    }
}